### PR TITLE
[WIP] Articulated Objects Support

### PR DESCRIPTION
**Pull request by aclegg3**
## Motivation and Context

WIP PR to track remaining work on merging ArticulatedObject support into Habitat-sim master branch. 

Feature TODO:
- URDF support in SceneInstanceAttributes structure
- ManagedArticulatedObject wrapper API and refactor
- Visual primitive shapes from URDF
- Attaching sensors to articulated agents (blocked by #1151 )
- Query Link maximal velocity

## How Has This Been Tested

`examples/tutorials/URDF_robotics_tutorial.py` demonstrates and applies new feature set.

TODO: better testing of individual features.

## Types of changes

- [ ] Docs change / refactoring / dependency upgrade
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

### CODE REVIEW

**Comment by jturner65** [comment]
How about we make a separate PR to merge all the URDF test assets into master now?  That'd take care of 37 files.

File: examples/tutorials/physics_benchmarking.py [hunk 1]
```diff
@@ -1,4 +1,8 @@
+# Copyright (c) Facebook, Inc. and its affiliates.
```
Comment (jturner65) [inline examples/tutorials/physics_benchmarking.py:1]: The changes in this file can be in a separate PR, maybe?

File: src/esp/physics/bullet/BulletRigidStage.h [hunk 2]
```diff
@@ -56,6 +56,8 @@ class BulletRigidStage : public BulletBase, public RigidStage {
       const std::vector<assets::CollisionMeshData>& meshGroup,
       const assets::MeshTransformNode& node);
 
+  std::string getCollisionDebugName(int subpartId);
+
```
Comment (jturner65) [inline src/esp/physics/bullet/BulletRigidStage.h:60]: docstring

File: src/esp/physics/bullet/BulletPhysicsManager.h [hunk 3]
```diff
@@ -253,6 +442,10 @@ class BulletPhysicsManager : public PhysicsManager {
    */
   bool isMeshPrimitiveValid(const assets::CollisionMeshData& meshData) override;
 
+  void lookUpObjectIdAndLinkId(const btCollisionObject* colObj,
+                               int* objectId,
+                               int* linkId) const;
+
```
Comment (jturner65) [inline src/esp/physics/bullet/BulletPhysicsManager.h:448]: docstring

File: src/esp/physics/RigidObject.h [hunk 4]
```diff
@@ -132,6 +132,11 @@ class RigidObject : public RigidBase {
         metadata::attributes::ObjectAttributes>();
   };
 
+  virtual void deferUpdate() { isDeferringUpdate_ = true; }
+  virtual void updateNodes(CORRADE_UNUSED bool force = false) {
+    isDeferringUpdate_ = false;
+  }
+
```
Comment (jturner65) [inline src/esp/physics/RigidObject.h:139]: Should this be moved to a base class (RigidBase, or PhysicsObjectBase ?)

File: src/esp/physics/ArticulatedObject.h [hunk 5]
```diff
@@ -0,0 +1,525 @@
...
+
+////////////////////////////////////
+// Articulated Object
+////////////////////////////////////
+
+/**
+ * @brief An articulated rigid object (i.e. kinematic chain). Abstract class to
+ * be derived by physics simulator specific implementations.
+ */
+class ArticulatedObject : public Magnum::SceneGraph::AbstractFeature3D {
```
Comment (jturner65) [inline src/esp/physics/ArticulatedObject.h:221]: As we discussed, this should inherit from PhysicsObjectBase :).

File: src/esp/physics/bullet/BulletDebugManager.h [hunk 6]
```diff
@@ -0,0 +1,79 @@
+// Copyright (c) Facebook, Inc. and its affiliates.
+// This source code is licensed under the MIT license found in the
+// LICENSE file in the root directory of this source tree.
+
+#pragma once
```
Comment (jturner65) [inline src/esp/physics/bullet/BulletDebugManager.h:5]: header guard instead of pragma.

File: src/esp/physics/URDFImporter.h [hunk 7]
```diff
@@ -0,0 +1,156 @@
+// Copyright (c) Facebook, Inc. and its affiliates.
+// This source code is licensed under the MIT license found in the
+// LICENSE file in the root directory of this source tree.
+
+// Code adapted from Bullet3/examples/Importers/ImportURDFDemo ...
+
+#pragma once
```
Comment (jturner65) [inline src/esp/physics/URDFImporter.h:7]: header guard instead of pragma.

File: src/esp/io/URDFParser.h [hunk 8]
```diff
@@ -0,0 +1,386 @@
+// Copyright (c) Facebook, Inc. and its affiliates.
+// This source code is licensed under the MIT license found in the
+// LICENSE file in the root directory of this source tree.
+
+// Code adapted from Bullet3/examples/Importers/ImportURDFDemo ...
+
+#pragma once
```
Comment (jturner65) [inline src/esp/io/URDFParser.h:7]: header guard instead of pragma.

File: src/esp/physics/ArticulatedObject.h [hunk 9]
```diff
@@ -0,0 +1,525 @@
+// Copyright (c) Facebook, Inc. and its affiliates.
+// This source code is licensed under the MIT license found in the
+// LICENSE file in the root directory of this source tree.
+
+#pragma once
```
Comment (jturner65) [inline src/esp/physics/ArticulatedObject.h:5]: header guard instead of pragma.

File: src/esp/physics/CollisionGroupHelper.h [hunk 10]
```diff
@@ -0,0 +1,44 @@
+// Copyright (c) Facebook, Inc. and its affiliates.
+// This source code is licensed under the MIT license found in the
+// LICENSE file in the root directory of this source tree.
+
+#pragma once
```
Comment (jturner65) [inline src/esp/physics/CollisionGroupHelper.h:5]: header guard instead of pragma.

**Comment by aclegg3** [comment]
Replaced by #1273.
